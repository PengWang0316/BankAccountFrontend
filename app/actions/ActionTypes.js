export const USER_LOGIN_SUCCESS = 'userLoginSuccess';
export const USER_LOGOUT_SUCCESS = 'userLogoutSuccess';
export const FETCH_ALL_ACCOUNT_SUCCESS = 'fetchAllAccountSuccess';
