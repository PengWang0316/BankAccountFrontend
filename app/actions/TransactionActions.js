import axios from 'axios';

import { FETCH_ALL_TRANSACTION_API } from '../config';

export const fetchAllTransaction = async (id) => {
  const { data } = await axios.get(FETCH_ALL_TRANSACTION_API, { params: { id } });
  return data.sort((pre, next) => new Date(next.date) - new Date(pre.date));
};

export default fetchAllTransaction;
