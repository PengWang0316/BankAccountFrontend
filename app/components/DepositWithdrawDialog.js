import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import I18n from '@kevinwang0316/i18n';

import {
  deposit as depositAction, withdraw as withdrawAction,
} from '../actions/AccountActions';

export const DepositWithdrawDialog = ({
  open, handleClose, isDeposit, accountId, deposit, withdraw,
}) => {
  const [amount, setAmount] = useState(0);

  const handleDeposit = () => deposit(amount, accountId);

  const handleWithdraw = () => withdraw(amount, accountId);

  const handleConfirmClick = () => {
    if (isDeposit) handleDeposit();
    else handleWithdraw();
    setAmount(0);
    handleClose();
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">{I18n.get(isDeposit ? 'depositDialogTitle' : 'withdrawDialogTitle')}</DialogTitle>
      <DialogContent>
        <DialogContentText>
          {`${isDeposit ? I18n.get('depositDialogText') : I18n.get('withdrawDialogText')} ${accountId}`}
        </DialogContentText>
        <TextField
          autoFocus
          margin="dense"
          id="amount"
          label={I18n.get('thAmount')}
          type="number"
          fullWidth
          value={amount}
          onChange={({ target }) => setAmount(target.value)}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          {I18n.get('cancel')}
        </Button>
        <Button onClick={handleConfirmClick} color="primary">
          {I18n.get(isDeposit ? 'deposit' : 'withdraw')}
        </Button>
      </DialogActions>
    </Dialog>
  );
};
DepositWithdrawDialog.propTypes = {
  open: PropTypes.bool,
  handleClose: PropTypes.func.isRequired,
  deposit: PropTypes.func.isRequired,
  withdraw: PropTypes.func.isRequired,
  isDeposit: PropTypes.bool.isRequired,
  accountId: PropTypes.string.isRequired,
};
DepositWithdrawDialog.defaultProps = { open: false };

/* istanbul ignore next */
const mapDispatchToProps = {
  withdraw: withdrawAction,
  deposit: depositAction,
};
export default connect(null, mapDispatchToProps)(DepositWithdrawDialog);
