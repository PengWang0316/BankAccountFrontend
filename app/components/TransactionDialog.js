import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import I18n from '@kevinwang0316/i18n';

import LoadingAnimation from './SharedComponents/LoadingAnimation';

/* istanbul ignore next */
const styles = theme => ({
  paper: {
    width: '90%',
    marginTop: theme.spacing.unit * 3,
    overflowX: 'auto',
    margin: 'auto',
    // minHeight: 400,
  },
  table: {
    minWidth: 800,
    // minHeight: 400,
  },
});

export const TransactionDialog = ({
  open, handleClose, transactions, classes, accountId,
}) => (
  <Dialog
    open={open}
    onClose={handleClose}
    aria-labelledby="form-dialog-title"
    maxWidth="md"
  >
    <DialogTitle id="form-dialog-title">{I18n.get('transicationDialogTitle')}</DialogTitle>
    <DialogContent>
      <Paper className={classes.paper}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell padding="dense">{I18n.get('thAccountId')}</TableCell>
              <TableCell padding="dense">{I18n.get('thType')}</TableCell>
              <TableCell padding="dense">{I18n.get('thAmount')}</TableCell>
              <TableCell align="justify" padding="dense">{I18n.get('thDate')}</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {!transactions && (
              <TableRow><LoadingAnimation /></TableRow>
            )}
            {transactions && transactions.map(({
              key, amount, type, date,
            }) => (
              <TableRow key={key}>
                <TableCell padding="dense">{accountId}</TableCell>
                <TableCell padding="dense">{type}</TableCell>
                <TableCell padding="dense">${amount}</TableCell>
                <TableCell padding="dense">{date}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </Paper>
    </DialogContent>
    <DialogActions>
      <Button onClick={handleClose} color="primary">
        {I18n.get('close')}
      </Button>
    </DialogActions>
  </Dialog>
);
TransactionDialog.propTypes = {
  open: PropTypes.bool,
  handleClose: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired,
  transactions: PropTypes.arrayOf(PropTypes.object),
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};
TransactionDialog.defaultProps = { open: false, transactions: null };

export default withStyles(styles)(TransactionDialog);
