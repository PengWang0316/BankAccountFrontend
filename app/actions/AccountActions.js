import axios from 'axios';

import { FETCH_ALL_ACCOUNT_SUCCESS } from './ActionTypes';
import { FETCH_ALL_ACCOUNT_API } from '../config';

const fetchAllAccountSuccess = accounts => ({
  type: FETCH_ALL_ACCOUNT_SUCCESS,
  accounts,
});

export const fetchAllAcount = () => dispatch => axios.get(FETCH_ALL_ACCOUNT_API)
  .then(({ data }) => dispatch(fetchAllAccountSuccess(data)));

export default fetchAllAcount;
