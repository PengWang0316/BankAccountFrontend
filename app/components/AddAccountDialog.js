import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import I18n from '@kevinwang0316/i18n';

import { addAccount as addAccountAction } from '../actions/AccountActions';

export const AddAccountDialog = ({ open, handleClose, addAccount }) => {
  const [name, setName] = useState('');
  const [balance, setBalance] = useState(0);

  const handleAddAccount = () => {
    addAccount({ name, balance });
    setName('');
    setBalance(0);
    handleClose();
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">{I18n.get('addAccountDialogTitle')}</DialogTitle>
      <DialogContent>
        <TextField
          autoFocus
          margin="dense"
          id="name"
          label="Name"
          type="text"
          fullWidth
          value={name}
          onChange={({ target }) => setName(target.value)}
        />
        <TextField
          margin="dense"
          id="balance"
          label="Balance"
          type="number"
          fullWidth
          value={balance}
          onChange={({ target }) => setBalance(target.value)}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          {I18n.get('cancel')}
        </Button>
        <Button onClick={handleAddAccount} color="primary">
          {I18n.get('add')}
        </Button>
      </DialogActions>
    </Dialog>
  );
};
AddAccountDialog.propTypes = {
  open: PropTypes.bool,
  handleClose: PropTypes.func.isRequired,
  addAccount: PropTypes.func.isRequired,
};
AddAccountDialog.defaultProps = { open: false };

/* istanbul ignore next */
const mapDispatchToProps = { addAccount: addAccountAction };
export default connect(null, mapDispatchToProps)(AddAccountDialog);
