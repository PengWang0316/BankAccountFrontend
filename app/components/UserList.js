/* eslint-disable react/jsx-one-expression-per-line */
import React, { useEffect, useState, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import I18n from '@kevinwang0316/i18n';

import { fetchAllAccount as fetchAllAccountsAction } from '../actions/AccountActions';
import { fetchAllTransaction } from '../actions/TransactionActions';
import TransactionDialog from './TransactionDialog';
import DepositWithdrawDialog from './DepositWithdrawDialog';

/* istanbul ignore next */
const styles = theme => ({
  root: {
    width: '70%',
    marginTop: theme.spacing.unit * 3,
    overflowX: 'auto',
    margin: 'auto',
  },
  table: {
    minWidth: 700,
  },
  row: {
    '&:hover': {
      cursor: 'pointer',
    },
  },
});

export const UserList = ({ classes, accounts, fetchAllAccount }) => {
  const [openTransactionDialog, setOpenTransactionDialog] = useState(false);
  const [openDepositWithdrawDialog, setOpenDepositWithdrawDialog] = useState(false);
  const [isDeposit, setIsDeposit] = useState(true);
  const [transactions, setTransactions] = useState(null);
  const [accountId, setAccountId] = useState('');

  useEffect(() => {
    if (!accounts) fetchAllAccount();
  });

  const handleTransactionDialogClose = () => setOpenTransactionDialog(false);

  const handleRowClick = async ({ target: { parentNode: { id } } }) => {
    setTransactions(null);
    setAccountId(id);
    setOpenTransactionDialog(true);
    const result = await fetchAllTransaction(id);
    setTransactions(result);
  };

  const handleDepositClick = (event) => {
    event.stopPropagation();
    setAccountId(event.target.tagName === 'BUTTON' ? event.target.parentNode.parentNode.id : event.target.parentNode.parentNode.parentNode.id);
    setIsDeposit(true);
    setOpenDepositWithdrawDialog(true);
  };

  const handleWithdrawClick = (event) => {
    event.stopPropagation();
    setAccountId(event.target.tagName === 'BUTTON' ? event.target.parentNode.parentNode.id : event.target.parentNode.parentNode.parentNode.id);
    setIsDeposit(false);
    setOpenDepositWithdrawDialog(true);
  };

  const handleDWDialogClose = () => setOpenDepositWithdrawDialog(false);

  return (
    <Fragment>
      <Paper className={classes.root}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell>{I18n.get('thAccountId')}</TableCell>
              <TableCell>{I18n.get('thName')}</TableCell>
              <TableCell>{I18n.get('thBalance')}</TableCell>
              <TableCell align="right">{I18n.get('thActions')}</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {accounts && Object.keys(accounts).map(key => (
              <TableRow key={key} id={key} onClick={handleRowClick} className={classes.row}>
                <TableCell>{accounts[key].accountId}</TableCell>
                <TableCell>{accounts[key].name}</TableCell>
                <TableCell>${accounts[key].balance}</TableCell>
                <TableCell align="right">
                  <Button size="small" color="primary" onClick={handleDepositClick}>{I18n.get('deposit')}</Button>
                  <Button size="small" color="secondary" onClick={handleWithdrawClick}>{I18n.get('withdraw')}</Button>
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </Paper>
      <TransactionDialog
        open={openTransactionDialog}
        handleClose={handleTransactionDialogClose}
        transactions={transactions}
        accountId={accountId}
      />
      <DepositWithdrawDialog
        open={openDepositWithdrawDialog}
        handleClose={handleDWDialogClose}
        accountId={accountId}
        isDeposit={isDeposit}
      />
    </Fragment>
  );
};
UserList.propTypes = {
  accounts: PropTypes.shape({
    accountId: PropTypes.string,
    name: PropTypes.string,
    balance: PropTypes.number,
  }),
  fetchAllAccount: PropTypes.func.isRequired,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};
UserList.defaultProps = { accounts: null };

/* istanbul ignore next */
const mapStateToProps = ({ accounts }) => ({ accounts });
/* istanbul ignore next */
const mapDispatchToProps = { fetchAllAccount: fetchAllAccountsAction };
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(UserList));
