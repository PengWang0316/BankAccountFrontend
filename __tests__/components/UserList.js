import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { UserList } from '../../app/components/UserList';
import { fetchAllTransaction } from '../../app/actions/TransactionActions';

jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/TableCell', () => 'TableCell');
jest.mock('@material-ui/core/TableHead', () => 'TableHead');
jest.mock('@material-ui/core/TableRow', () => 'TableRow');
jest.mock('@material-ui/core/Paper', () => 'Paper');
jest.mock('../../app/components/TransactionDialog', () => 'TransactionDialog');
jest.mock('../../app/components/DepositWithdrawDialog', () => 'DepositWithdrawDialog');
jest.mock('@kevinwang0316/i18n', () => ({ get: key => key }));
jest.mock('../../app/actions/TransactionActions', () => ({ fetchAllTransaction: jest.fn() }));

describe('UserList', () => {
  const defaultProps = {
    classes: { root: 'root', table: 'table', row: 'row' },
    accounts: { id1: { accountid: 'id1', name: 'name1', balance: 100 }, id2: { accountid: 'id2', name: 'name2', balance: 130 } },
    fetchAllAccount: jest.fn(),
  };
  const getShallowComponent = (props = defaultProps) => shallow(<UserList {...props} />);

  beforeEach(() => jest.clearAllMocks());

  test('Initial with accounts', () => {
    renderer.create(<UserList {...defaultProps} />);
    expect(defaultProps.fetchAllAccount).not.toHaveBeenCalled();
  });

  test('Initial without accounts', () => {
    renderer.create(<UserList {...{ ...defaultProps, accounts: null }} />);
    expect(defaultProps.fetchAllAccount).toHaveBeenCalledTimes(1);
  });

  test('handleRowClick', () => {
    const component = getShallowComponent();
    const tableRows = component.find('TableRow');
    tableRows.at(1).simulate('click', { target: { parentNode: { id: 'id' } } });

    expect(fetchAllTransaction).toHaveBeenCalledTimes(1);
  });

  test('handleDepositClick', () => {
    const mockStopPropagationFn = jest.fn();
    const component = getShallowComponent();
    const buttons = component.find('Button');
    buttons.at(0).simulate('click', { stopPropagation: mockStopPropagationFn, target: { tagName: 'BUTTON', parentNode: { parentNode: { id: 'id1' } } } });

    expect(mockStopPropagationFn).toHaveBeenCalledTimes(1);

    buttons.at(0).simulate('click', { stopPropagation: mockStopPropagationFn, target: { tagName: 'SPAN', parentNode: { parentNode: { parentNode: { id: 'id1' } } } } });
    expect(mockStopPropagationFn).toHaveBeenCalledTimes(2);
  });

  test('handleWithdrawClick', () => {
    const mockStopPropagationFn = jest.fn();
    const component = getShallowComponent();
    const buttons = component.find('Button');
    buttons.at(1).simulate('click', { stopPropagation: mockStopPropagationFn, target: { tagName: 'BUTTON', parentNode: { parentNode: { id: 'id1' } } } });

    expect(mockStopPropagationFn).toHaveBeenCalledTimes(1);

    buttons.at(1).simulate('click', { stopPropagation: mockStopPropagationFn, target: { tagName: 'SPAN', parentNode: { parentNode: { parentNode: { id: 'id1' } } } } });
    expect(mockStopPropagationFn).toHaveBeenCalledTimes(2);
  });

  test('UserList with accounts snapshot', () => expect(renderer.create(<UserList {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('UserList without accounts snapshot', () => expect(renderer.create(<UserList {...{ ...defaultProps, accounts: null }} />).toJSON()).toMatchSnapshot());
});
