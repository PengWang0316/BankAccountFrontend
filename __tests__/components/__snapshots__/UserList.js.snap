// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList UserList with accounts snapshot 1`] = `
Array [
  <Paper
    className="root"
  >
    <Table
      className="table"
    >
      <TableHead>
        <TableRow>
          <TableCell>
            thAccountId
          </TableCell>
          <TableCell>
            thName
          </TableCell>
          <TableCell>
            thBalance
          </TableCell>
          <TableCell
            align="right"
          >
            thActions
          </TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow
          className="row"
          id="id1"
          onClick={[Function]}
        >
          <TableCell />
          <TableCell>
            name1
          </TableCell>
          <TableCell>
            $
            100
          </TableCell>
          <TableCell
            align="right"
          >
            <Button
              color="primary"
              onClick={[Function]}
              size="small"
            >
              deposit
            </Button>
            <Button
              color="secondary"
              onClick={[Function]}
              size="small"
            >
              withdraw
            </Button>
          </TableCell>
        </TableRow>
        <TableRow
          className="row"
          id="id2"
          onClick={[Function]}
        >
          <TableCell />
          <TableCell>
            name2
          </TableCell>
          <TableCell>
            $
            130
          </TableCell>
          <TableCell
            align="right"
          >
            <Button
              color="primary"
              onClick={[Function]}
              size="small"
            >
              deposit
            </Button>
            <Button
              color="secondary"
              onClick={[Function]}
              size="small"
            >
              withdraw
            </Button>
          </TableCell>
        </TableRow>
      </TableBody>
    </Table>
  </Paper>,
  <TransactionDialog
    accountId=""
    handleClose={[Function]}
    open={false}
    transactions={null}
  />,
  <DepositWithdrawDialog
    accountId=""
    handleClose={[Function]}
    isDeposit={true}
    open={false}
  />,
]
`;

exports[`UserList UserList without accounts snapshot 1`] = `
Array [
  <Paper
    className="root"
  >
    <Table
      className="table"
    >
      <TableHead>
        <TableRow>
          <TableCell>
            thAccountId
          </TableCell>
          <TableCell>
            thName
          </TableCell>
          <TableCell>
            thBalance
          </TableCell>
          <TableCell
            align="right"
          >
            thActions
          </TableCell>
        </TableRow>
      </TableHead>
      <TableBody />
    </Table>
  </Paper>,
  <TransactionDialog
    accountId=""
    handleClose={[Function]}
    open={false}
    transactions={null}
  />,
  <DepositWithdrawDialog
    accountId=""
    handleClose={[Function]}
    isDeposit={true}
    open={false}
  />,
]
`;
